### PR TITLE
feat(execution-engine)!: fail :error: now bubbles the original error up [fixes VM-342]

### PR DESCRIPTION
**Pull request by raftedproc**
feat(execution-engine)!: fail :error: now bubbles the original error up [fixes VM-342]

### CODE REVIEW

File: air/src/execution_step/instructions/xor.rs [hunk 1]
```diff
@@ -34,11 +34,14 @@ impl<'i> super::ExecutableInstruction<'i> for Xor<'i> {
...
 
+                exec_ctx.error_descriptor.set_original_execution_error(&e);
+                exec_ctx.error_descriptor.enable_error_setting();
+
                 let right_subgraph_result = self.1.execute(exec_ctx, trace_ctx);
                 // This sets :error: to a no-error state.
                 // Please note the right_subgraph_result might be an Error that bubbles up to an :error:
                 // above this execute().
-                exec_ctx.error_descriptor.clear_error();
+                exec_ctx.error_descriptor.clear_error_object_if_needed();
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/xor.rs:44]: `:error:` must be cleaned every time after a right xor subgraph has executed

File: air/src/execution_step/instructions/fail.rs [hunk 2]
```diff
@@ -127,11 +128,21 @@ fn fail_with_error(exec_ctx: &mut ExecutionCtx<'_>) -> ExecutionResult<()> {
...
-    fail_with_error_object(exec_ctx, error.clone(), tetraplet.clone(), provenance.clone())
+    let result = match orig_catchable {
+        Some(orig_catchable) => {
+            let catchable_to_return = orig_catchable.clone();
+            let _ = fail_with_error_object(exec_ctx, error.clone(), tetraplet.clone(), provenance.clone());
+            Err(ExecutionError::Catchable(catchable_to_return.into()))
+        }
+        None => fail_with_error_object(exec_ctx, error.clone(), tetraplet.clone(), provenance.clone()),
+    };
+    exec_ctx.error_descriptor.disable_error_setting();
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/fail.rs:144]: Setting error should be disabled for all error variants (e.g., for an error from a scalar).

File: air/src/execution_step/instructions/mod.rs [hunk 3]
```diff
@@ -57,7 +57,10 @@ macro_rules! execute {
                 $exec_ctx.set_errors(&e, &$instr.to_string(), None, $instr.log_errors_with_peer_id());
                 Err(e)
             }
-            v => v,
+            v => {
+                $exec_ctx.error_descriptor.enable_error_setting();
```
Comment (mikevoronov) [inline air/src/execution_step/instructions/mod.rs:61]: Why do you need it?

File: air/src/execution_step/execution_context/context.rs [hunk 4]
```diff
@@ -193,8 +193,12 @@ impl ExecutionCtx<'_> {
             None
         };
 
-        self.error_descriptor
-            .try_to_set_error_from_exec_error(error, instruction, peer_id, tetraplet.clone());
+        if self.error_descriptor.error_setting_is_enabled() {
```
Comment (mikevoronov) [inline air/src/execution_step/execution_context/context.rs:196]: it's better to move this check inside the `try_to_set_error_from_exec_error` function.

File: air/src/execution_step/instructions/mod.rs (hunk 3 above)
Comment (raftedproc) [reply air/src/execution_step/instructions/mod.rs:61]: How would you re-enable errors for the second call in this AIR?
```
(seq
    (xor
       (xor
          (call to fail 1)
          (fail :error:)
       )
       (none)
     )
  (call to fail 2)
)
```

Comment (raftedproc) [reply air/src/execution_step/instructions/mod.rs:61]: Moved into Xor::execute

File: air/src/execution_step/execution_context/context.rs (hunk 4 above)
Comment (raftedproc) [reply air/src/execution_step/execution_context/context.rs:196]: Done

File: air/src/execution_step/execution_context/instruction_error/error_descriptor.rs [hunk 5]
```diff
@@ -34,6 +36,11 @@ impl ErrorDescriptor {
     ) {
         use super::get_instruction_error_from_exec_error;
 
+        // Returns early if there is an error to bubble up.
```
Comment (mikevoronov) [inline air/src/execution_step/execution_context/instruction_error/error_descriptor.rs:39]:
```suggestion
        // returns early if there is an error to bubble up
```

File: air/src/execution_step/execution_context/instruction_error/error_descriptor.rs [hunk 6]
```diff
@@ -34,6 +36,11 @@ impl ErrorDescriptor {
...
+        if !self.error_can_be_set {
```
Comment (mikevoronov) [inline air/src/execution_step/execution_context/instruction_error/error_descriptor.rs:40]: let's write these two conditions in one if block

**Review by mikevoronov — APPROVED** [review]
utACK

File: air/src/execution_step/execution_context/instruction_error/error_descriptor.rs (hunk 6 above)
Comment (raftedproc) [reply air/src/execution_step/execution_context/instruction_error/error_descriptor.rs:40]: Fixed

File: air/src/execution_step/execution_context/instruction_error/error_descriptor.rs (hunk 5 above)
Comment (raftedproc) [reply air/src/execution_step/execution_context/instruction_error/error_descriptor.rs:39]: Fixed

**Pull request merged**